x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/SearchAfterDocumentsIterator.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ml.utils.persistence;
+
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.sort.FieldSortBuilder;
+import org.elasticsearch.xpack.ml.utils.MlIndicesUtils;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+/**
+ * An iterator useful to fetch a large number of documents of type T
+ * and iterate through them in batches of 10,000.
+ *
+ * In terms of functionality this is very similar to {@link BatchedDocumentsIterator}
+ * the difference being that this uses search after rather than scroll.
+ *
+ * Search after has the advantage that the scroll context does not have to be kept
+ * alive so if processing each batch takes a long time search after should be
+ * preferred to scroll.
+ */
+public abstract class SearchAfterDocumentsIterator<T> implements BatchedIterator<T> {
+
+    private static final int BATCH_SIZE = 10_000;
+
+    private final OriginSettingClient client;
+    private final String index;
+    private volatile long count;
+    private volatile long totalHits;
+
+    protected SearchAfterDocumentsIterator(OriginSettingClient client, String index) {
+        this.client = Objects.requireNonNull(client);
+        this.index = Objects.requireNonNull(index);
+        this.totalHits = -1;
+        this.count = 0;
+    }
+
+    /**
+     * Returns {@code true} if the iteration has more elements or
+     * no searches have been been run and it is unknown if there is a next.
+     *
+     * @return {@code true} if the iteration has more elements or the first
+     * search has not been run
+     */
+    @Override
+    public boolean hasNext() {
+        return count != totalHits;
```

Comment:
Does search after protect us from new docs being added?

It seems to me that there is a possibility of `count > totalHits` if new documents are added + the index is refreshed.

It might be good to have this as `count <= totalHits` as adjust the initial values + initialization of values accordingly.

Reply:
That is a very good point and with deletes `count` may never `==  totalHits`. 

I changed the logic so that hasNext will return false only if the last search returned less than the requested number for hits e.g. I asked for 10 results but only got 8 back so there are no more search results and that is the end of the iteration.